pkg/raft/raft.go

```diff
@@ -1342,7 +1342,38 @@ func (r *raft) becomeLeader() {
 		panic("invalid transition [follower -> leader]")
 	}
 	r.step = stepLeader
-	r.reset(r.Term, nil)
+
+	// A special visit function to use the recorded hints and optimize the first
+	// round of MsgApp after becoming leader.
+	r.reset(r.Term, func(id pb.PeerID, pr *tracker.Progress) {
+		matchGuess := tracker.MatchGuess{Index: r.raftLog.lastIndex()}
+		if guess, ok := r.electionTracker.MatchGuess(id); ok {
+			matchGuess = guess
+		}
+
+		*pr = tracker.Progress{
+			Match:       0,
+			MatchCommit: 0,
+			Next:        matchGuess.Index + 1,
+			Inflights:   tracker.NewInflights(r.maxInflight, r.maxInflightBytes),
+			IsLearner:   pr.IsLearner,
+		}
+
+		if matchGuess.Match {
+			// Follower's log matches the leader's log at next-1.
+			// Set the follower's state to StateReplicate early.
+			r.becomeReplicate(pr)
+			// NB: this is a bit of a hack, but we need to set the Next again since
+			// r.becomeReplicate sets pr.Next to match + 1.
+			// Where match is 0 in this case.
+			pr.Next = matchGuess.Index + 1
+		}
+
+		if pr.Next <= r.raftLog.storage.Compacted() {
+			pr.RecentActive = true
+		}
```

Comment:
As discussed in Slack, we can make this less corner-casey by saying that all peers with a response become `RecentActive`. This is consistent with the logic of setting `RecentActive` to true on all `Resp` messages.

E.g. a simple fix:

```golang
matchGuess := tracker.MatchGuess{Index: r.raftLog.lastIndex()}
replied := false
if guess, ok := r.electionTracker.MatchGuess(id); ok {
	matchGuess, replied = guess, true
}

*pr = Progress {
	...
	RecentActive: replied,
}
```